### PR TITLE
Add SV count outlier cutoff table to Module03 and bug fix for MasterVcfQC

### PR DESCRIPTION
**Pull request by mwalker174**
This is a rebase of #65 with some cleanup and updated dockers. 

Type- and algorithm-specific SV count ranges can now be specified in a table for filtering outlier samples in Module03 (in addition to the current IQR cutoffs). 

Also includes a bug fix to `analyze_fams.R` in MasterVcfQc.

### CODE REVIEW

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  for(n in c(0:x_max)){
+    abline(h=quantil_list[4]+n*iqr, col='grey', lty=2)
+    abline(h=quantil_list[2]-n*iqr, col='grey', lty=2)
+  }
+  
+  sample_counts = data.frame('iqr'=0,'sample_counts'=0)
+  rec=0
+  for(j in c(1:x_max)){
+    cff = c(quantil_list[2]-j*iqr, quantil_list[4]+j*iqr)
+    tmp2=tmp[tmp$count<cff[2] & tmp$count>cff[1],]
```
Comment (cwhelan) [inline src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:59]: `tmp2` -> `counts_outside_cutoffs`?

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            border=adjustcolor("darkgreen",alpha=0.3),col=NA, yaxt='n', lwd=2)
+    points(df_x + j-1,df_y,pch=21,lwd=1.5,cex=0.1,col="darkgreen")
+    
+    rec=rec+1
+    sample_counts[rec,1]=j
+    sample_counts[rec,2]=nrow(tmp2)
+  }
+  
+  par(fig=c(0,1,.3,.6), new=T)
+  tmp = all_counts[all_counts[,4]==i,]
```
Comment (cwhelan) [inline src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:76]: `tmp` -> `counts_for_svtype`

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+opts <- args$options
+dat=read.table(opts$input, header = T)
+all_counts=read.table(opts$svcount)
+all_counts=all_counts[all_counts[,1]>0,]
+output_prefix = gsub('.txt','',opts$input)
+
+svtype=unique(dat$svtype)
+x_max= opts$max
+for(i in svtype){
```
Comment (cwhelan) [inline src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:37]: `i` -> `svtype`

File: src/sv-pipeline/scripts/vcf_qc/analyze_fams.R [hunk 4]
```diff
@@ -125,12 +125,17 @@ getFamDat <- function(dat,proband,father=NA,mother=NA,biallelic=T,nocall.placeho
...
-    any(as.numeric(vals)==nocall.placeholder)
-  }))
-  if(length(exclude) > 0){
-    vlist <- vlist[-exclude,]
+    which(as.numeric(vals)==nocall.placeholder)
+  })
+
+  exclude_out = c()
+  for(exclude_x in exclude){  exclude_out=c(exclude_out,exclude_x) }
+  exclude_out=unique(exclude_out)
```
Comment (cwhelan) [inline src/sv-pipeline/scripts/vcf_qc/analyze_fams.R:135]: Can't this block just be written as `exclude_out <- unique(exclude)`?

File: src/sv-pipeline/scripts/vcf_qc/analyze_fams.R [hunk 5]
```diff
@@ -1339,8 +1344,8 @@ if(nrow(trios)>0){
   #Read data
   trio.dat <- apply(trios[,2:4],1,function(IDs){
     IDs <- as.character(IDs)
-    return(getFamDat(dat=dat,proband=IDs[1],father=IDs[2],
-                     mother=IDs[3],biallelic=!multiallelics))
+    print(IDs)
```
Comment (cwhelan) [inline src/sv-pipeline/scripts/vcf_qc/analyze_fams.R:1347]: Not sure if you want to keep this `print` call in or if it was just for debugging.

**Review by cwhelan — CHANGES_REQUESTED** [review]
This generally looks fine to me logic-wise, but I couldn't help requesting some formatting/style/readability changes in the R scripts.

File: src/sv-pipeline/scripts/vcf_qc/analyze_fams.R (hunk 4 above)
Comment (xuefzhao) [reply src/sv-pipeline/scripts/vcf_qc/analyze_fams.R:135]: no, this code is to transfer data formats

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R (hunk 3 above)
Comment (xuefzhao) [reply src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:37]: that'll be a bug and overwrite the original `svtype`

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R (hunk 2 above)
Comment (xuefzhao) [reply src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:76]: fixed

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R (hunk 1 above)
Comment (xuefzhao) [reply src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:59]: fixed

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R [hunk 6]
```diff
@@ -0,0 +1,89 @@
+getJitter <- function(x,dens){
+  jit <- dens$y[head(which(dens$x>=x),1)]
+  return(jit)
+}
+jitterX <- function(y,dens){
+  return(jitter(1,amount=getJitter(y,dens)))
+}
+
+
+library(optparse)
```
[the remark this answers is not in the document]
Comment (xuefzhao) [reply src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:10]: fixed

File: src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+cffs <-read.table(CutoffFile, header=T, sep='\t')
+# Determine outlier samples
+fails <- unique(as.character(unlist(sapply(unique(counts$svtype),function(svtype){
+  # Get counts corresponding to svtype
+  vals <- counts$count[which(counts$svtype==svtype)]
+  
+  # Determine cutoff
+  quartiles <- as.numeric(quantile(vals,probs=c(0.25,0.75),na.rm=T))
+
+  cutoffs <- c(cffs[cffs$algorithm==algorithm & cffs$svtype == svtype,c(3,4)])
```
[the remark this answers is not in the document]
Comment (xuefzhao) [reply src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R:31]: fixed

File: src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+# Read command line arguments
+options(stringsAsFactors=F)
+args <- commandArgs(trailingOnly=TRUE)
+if(length(args)!=4){
+  stop("Four positional arguments expected: svcounts.txt, CutoffFile, OUTFILE and algorithm")
+}
+countsfile <- as.character(args[1])
+CutoffFile<- as.character(args[2])
+OUTFILE <- as.character(args[3])
```
[the remark this answers is not in the document]
Comment (xuefzhao) [reply src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R:17]: fixed

File: src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+# Contact: Ryan Collins <rlcollins@g.harvard.edu>
+# Distributed under terms of the MIT license.
+
+# Helper script to return list of outlier samples from an svcounts.txt file
+
+# Read command line arguments
+options(stringsAsFactors=F)
+args <- commandArgs(trailingOnly=TRUE)
+if(length(args)!=4){
+  stop("Four positional arguments expected: svcounts.txt, CutoffFile, OUTFILE and algorithm")
```
[the remark this answers is not in the document]
Comment (xuefzhao) [reply src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R:13]: fixed

File: src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+countsfile <- as.character(args[1])
+CutoffFile<- as.character(args[2])
+OUTFILE <- as.character(args[3])
+algorithm <- as.character(args[4])
+
+# Read counts
+counts <- read.table(countsfile,header=T,sep="\t")
+cffs <-read.table(CutoffFile, header=T, sep='\t')
+# Determine outlier samples
+fails <- unique(as.character(unlist(sapply(unique(counts$svtype),function(svtype){
```
[the remark this answers is not in the document]
Comment (xuefzhao) [reply src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R:24]: fixed

File: src/sv-pipeline/scripts/vcf_qc/analyze_fams.R (hunk 5 above)
Comment (xuefzhao) [reply src/sv-pipeline/scripts/vcf_qc/analyze_fams.R:1347]: fixed

**Comment by xuefzhao** [comment]
Thanks for comments! all addressed.

File: src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R [hunk 11]
```diff
@@ -5,36 +5,42 @@
...
-                      & (counts$count<cutoffs[1] | counts$count>cutoffs[2]))]
-  return(fails)
-}))))
-
-#Write list of fail samples to OUTFILE
-write.table(fails,OUTFILE,col.names=F,row.names=F,quote=F)
+OutFile <- as.character(args[3])
+Algorithm <- as.character(args[4])
+
+main<-function(CountsFile, CutoffFile,OutFile,  Algorithm){
```
Comment (cwhelan) [inline src/sv-pipeline/03_variant_filtering/scripts/get_outliers_from_svcounts.helper_V2.R:23]: Oh, when I said to name the function I meant the unnamed function applied in this loop, not to create a `main` function. Not a big deal though.

File: src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R (hunk 3 above)
Comment (cwhelan) [reply src/sv-pipeline/scripts/downstream_analysis_and_filtering/plot_svcount_outliers.R:37]: Sorry, I guess I meant something like, rename `svtype` to `svtypes` and then rename `i` to `svtype`. But again, not really a big deal, just trying to improve readability.

**Review by cwhelan — APPROVED** [review]
Thanks for addressing changes.

**Pull request merged**